LibraBFT/ImplShared/Consensus/Types/EpochIndep.agda

```diff
@@ -476,34 +476,86 @@ module LibraBFT.ImplShared.Consensus.Types.EpochIndep where
+    constructor Timeout∙new
+    field
+      _toEpoch : Epoch
+      _toRound : Round
+  open Timeout public
+  unquoteDecl toEpoch   toRound = mkLens (quote Timeout)
+             (toEpoch ∷ toRound ∷ [])
+  postulate instance enc-Timeout : Encoder Timeout
+
+  record TimeoutCertificate : Set where
+    constructor mkTimeoutCertificate
+    field
+      _tcTimeout    : Timeout
+      _tcSignatures : KVMap Author Signature
+  open TimeoutCertificate public
+  unquoteDecl tcTimeout   tcSignatures = mkLens (quote TimeoutCertificate)
+             (tcTimeout ∷ tcSignatures ∷ [])
+
+  TimeoutCertificate∙new : Timeout → TimeoutCertificate
+  TimeoutCertificate∙new to = mkTimeoutCertificate to Map.empty
+
+  -- IMPL-DIFF : only a getter in haskell
+  tcEpoch : Lens TimeoutCertificate Epoch
+  tcEpoch = tcTimeout ∙ toEpoch
+
+  -- IMPL-DIFF : only a getter in haskell
+  tcRound : Lens TimeoutCertificate Round
+  tcRound = tcTimeout ∙ toRound
+
   record SyncInfo : Set where
     constructor mkSyncInfo -- Bare constructor to enable pattern matching against SyncInfo; "smart"
                            -- constructor SyncInfo∙new is below
     field
       _siHighestQuorumCert  : QuorumCert
       _sixxxHighestCommitCert  : Maybe QuorumCert
-      -- _siHighestTimeoutCert : Maybe TimeoutCert -- Not used yet.
+      _siHighestTimeoutCert : Maybe TimeoutCertificate
   open SyncInfo public
   -- Note that we do not automatically derive a lens for siHighestCommitCert;
   -- it is defined manually below.
-  unquoteDecl siHighestQuorumCert   sixxxHighestCommitCert = mkLens (quote SyncInfo)
-             (siHighestQuorumCert ∷ sixxxHighestCommitCert ∷ [])
+  unquoteDecl siHighestQuorumCert   sixxxHighestCommitCert   siHighestTimeoutCert = mkLens (quote SyncInfo)
+             (siHighestQuorumCert ∷ sixxxHighestCommitCert ∷ siHighestTimeoutCert ∷ [])
   postulate instance enc-SyncInfo : Encoder SyncInfo
 
-  SyncInfo∙new : QuorumCert → QuorumCert → SyncInfo
-  SyncInfo∙new highestQuorumCert highestCommitCert =
+  SyncInfo∙new : QuorumCert → QuorumCert → Maybe TimeoutCertificate → SyncInfo
+  SyncInfo∙new highestQuorumCert highestCommitCert highestTimeoutCert =
     record { _siHighestQuorumCert    = highestQuorumCert
            ; _sixxxHighestCommitCert = if highestQuorumCert QCBoolEq highestCommitCert
-                                       then nothing else (just highestCommitCert) }
+                                       then nothing else (just highestCommitCert)
+           ; _siHighestTimeoutCert   = highestTimeoutCert }
 
+  -- getter only in Haskell
   siHighestCommitCert : Lens SyncInfo QuorumCert
   siHighestCommitCert =
     mkLens' (λ x → fromMaybe (x ^∙ siHighestQuorumCert) (x ^∙ sixxxHighestCommitCert))
-            (λ x si → record x { _sixxxHighestCommitCert = just si })
+            (λ x qc → record x { _sixxxHighestCommitCert = just qc })
+
+  -- getter only in Haskell
+  siHighestCertifiedRound : Lens SyncInfo Round
```

Comment:
> We seem to be using a couple of different approaches for cases where we have only `Getter`s in the Haskell code. One, like here, is to define a `Lens` that doesn't obey the Lens Laws, which we can then use nicely as a `Getter` (which obviously meets our needs), and the other is to define functions that don't even pretend to be Lenses, and then change the usage so that it differs from the Haskell

I am in favour of writing all getter-onlys as "weird" lenses for now, since that not only makes the Agda implementation look more like the Haskell now, but should also make adding a proper `Getter` easier (since in theory we will not have to change the implementation code, just the types of the LBFT and lens operations).

Reply:
Yes, my thoughts exactly @cwjnkins.  @haroldcarr, does this make sense and if so, will you action this?

Reply:
@cwjnkins @mark-moir Yes - making "getters weird" (AGAIN) is fine with me.   I think that should be done in a separate PR (which I can do if you agree).